File: CivitasJava/src/Civitas/Especulador.java

package Civitas;
public class Especulador extends Jugador {
    private int fianza;
    private static int FactorEspeculador = 2;
    
    Especulador(Jugador jugador,int fianza){
        super(jugador);
        this.fianza = fianza;
        for(int i=0;i<super.getPropiedades().size();i++){
            super.getPropiedades().get(i).actualizaPropietarioPorConversion(this);
        }
        super.especulador = true;
    }
    
    boolean pagarFianza(){
        return super.getSaldo() > this.fianza;
    }
    
    public int getFianza(){
        return this.fianza;
    }
    
    protected static int getCasasMax() {
        return CasasMax;
    }
    
    protected static int getHotelesMax() {
        return HotelesMax;
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
        if(this.encarcelado)
            return false;
        else
            return this.paga(cantidad/2);
    }
    
    @Override
    protected boolean debeSerEncarcelado(){
        if (this.encarcelado)
            return false;
        else if(super.tieneSalvoconducto()){
            super.perderSalvoConducto();
            Diario.getInstance().ocurreEvento("Jugador " + this.nombre + 
                    " usa salvoconducto, no entra en la cárcel");
            return false;
        }
        else if(!super.tieneSalvoconducto() && this.pagarFianza()){
            super.modificarSaldo(-(this.fianza));
            Diario.getInstance().ocurreEvento("Jugador " + this.nombre + 
                    " paga con la fianza, no entra en la cárcel");
            return false;
        }
        else
            return true;
    }
    
    @Override
    public boolean puedoEdificarCasa(TituloPropiedad propiedad){
        return super.puedoGastar(propiedad.getPrecioEdificar()) && propiedad.getNumCasas() < getCasasMax()*FactorEspeculador;
    }
    
    @Override
    public boolean puedoEdificarHotel(TituloPropiedad propiedad){
        return this.puedoGastar(propiedad.getPrecioEdificar()) && propiedad.getNumHoteles() < getHotelesMax()*FactorEspeculador 
                && propiedad.getNumCasas()>=getCasasPorHotel();
    }
    
    
    @Override
    public String toString() {
        return "Jugador{" + "\n Nombre: " + super.getNombre() + ", \n numCasillaActual: " + 
                super.getNumCasillaActual() + ", \n PuedeComprar: " + super.getPuedeComprar() + ", \n Saldo=" + 
                super.getSaldo() + ", \n Encarcelado: " + super.isEncarcelado() + " \n Propiedades: " + 
                super.getPropiedades() + " \n Fianza: " + this.getFianza() + "\n}";
    }
    
    
    
}

File: CivitasRuby/lib/especulador.rb
# encoding:utf-8
require_relative 'jugador.rb'
module Civitas
  class Especulador < Jugador
    
    attr_accessor :fianza
    @@FACTOR_ESPECULADOR = 2
    
    def self.nuevo_especulador(jugador,fianza)
      @fianza
      especulador = super(jugador)
      especulador.fianza = fianza
      for i in especulador.propiedades
        i.actualiza_propietario_por_conversion(self)
      end
      
      return especulador
    end
    
    def pagar_fianza
      return saldo > @fianza
    end
    
        
    def self.getCasas_Max
      return @@CASAS_MAX
    end
    
    def self.getHoteles_Max
      return @@HOTELES_MAX
    end
    
    def paga_impuesto(cantidad)
      if super.encarcelado
        return false
      else
        return paga(cantidad/2)
      end
    end
    
    def debe_ser_encarcelado
      
      case (encarcelado)
        when true
          return false
      when false
        if tiene_salvoconducto
          perder_salvoconducto
          Diario.instance.ocurre_evento("Jugador " + @nombre + 
               " tiene salvoconducto, no entra en la carcel \n")
          return false
        else
          if !tiene_salvoconducto && pagar_fianza
            modificar_saldo(-@fianza)
            Diario.instance.ocurre_evento("Jugador " + @nombre + 
               " paga con la fianza, no entra en la cárcel \n")
            return false
          else
            return true
          end
        end
      end
    end
    
    
    
    def puedo_edificar_casa(propiedad)
      return puedo_gastar(propiedad.precio_edificar) && propiedad.num_casas < Especulador::getCasas_Max * @@FACTOR_ESPECULADOR
    end
    
    def puedo_edificar_hotel(propiedad)
      return puedo_gastar(propiedad.precio_edificar) && 
        propiedad.num_hoteles < Especulador::getHoteles_Max * @@FACTOR_ESPECULADOR && 
        propiedad.num_casas >= @@CASAS_POR_HOTEL
    end
    
    
    def to_s
      "Jugador: { \n Nombre: #{@nombre} \n Encarcelado: #{@encarcelado} 
 Propiedades: #{@propiedades} \n Saldo: #{@saldo} \n Puede comprar: #{@puede_comprar} 
 Casilla actual: #{@num_casilla_actual} \n Salvoconducto: #{@salvoconducto} \n Fianza: #{@fianza} \n}"
    end
    
  end
end
